core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1437,4 +1487,42 @@ public void sync(Translog.Location location) {
         }
     }
 
+    private final AtomicBoolean asyncFlushRunning = new AtomicBoolean();
+
+    /**
+     * Schedules a flush if needed but won't schedule more than one flush concurrently. The flush will be executed on the
+     * Flush thread-pool asynchronously.
+     * @return <code>true</code> if a new flush is scheduled otherwise <code>false</code>.
+     */
+    public boolean maybeFlush() {
+        if (shouldFlush()) {
+            if (asyncFlushRunning.compareAndSet(false, true)) {
+                AbstractRunnable abstractRunnable = new AbstractRunnable() {
+                    @Override
+                    public void onFailure(Throwable t) {
+                        if (state != IndexShardState.CLOSED) {
+                            logger.warn("failed to flush index", t);
+                        }
+                    }
+                    @Override
+                    protected void doRun() throws Exception {
+                        flush(new FlushRequest());
+                    }
+
+                    @Override
+                    public void onAfter() {
+                        asyncFlushRunning.compareAndSet(true, false);
+                    }
+                };
+                if (shouldFlush()) {
```

Comment:
this double check seems redundant? we check it right before in the first line of the method? will simplify a bit the code, since there is no need for an `else`?